### ISSUE TITLE
Support ARM based Keycloak deployments on ROSA

### ISSUE THREAD

**Issue by kami619** [keycloak/keycloak-benchmark#1022]
### Description

- Identify remaining services on Keycloak Benchmark that don't support ARM, possibly leading to a mixed ARM and non-ARM cluster

**Comment by kami619** [comment]
- Minikube CI would fail to execute without an ARM runner for the build validation CI.

**Comment by kami619** [comment]
We found an issue with the custom image build we are performing and that might be defaulting to a specific architecture type of the imagestream and that needs to be addressed as specified in the article @ahus1 https://developers.redhat.com/articles/2024/08/27/building-multi-architecture-container-images-openshift-container-platform

**Comment by ahus1** [comment]
As discussed earlier today, supporting ARM on Minikube on GHI is out of scope for this. The main reason we're running Minikube is to test if our Helm charts are syntactically correct in our PRs. 

For the change we identified in that article - let's hope the fix is generic and that the changed version will work in both architectures.

**Comment by kami619** [comment]
A successful run: https://github.com/kami619/keycloak-benchmark/actions/runs/11616800201/job/32352177325

**Comment by kami619** [comment]
@ahus1 Also there is this question of implementing a conditional image build using a specific architecture https://github.com/keycloak/keycloak-benchmark/blob/fe770177c4ea24c0c07607b84fab0e8572f07a39/provision/keycloak-tasks/keycloak-image-helm/templates/keycloak-with-AWS-JDBC-wrapper.yaml#L24

and add the below snippet to the BuildConfig and implement a new Variable to control all these ?

```
nodeSelector:
    kubernetes.io/arch: amd64
```

Do we want to try and do it as part of this PR?

**Comment by ahus1** [comment]
> Do we want to try and do it as part of this PR?

The first PR shows that we are probably good to support a pure ARM cluster, and didn't find any blockers yet that would force us to run in a mixed cluster. 

So I'd say we proceed with minimal changes needed for a pure ARM cluster (which would then revert the scheduling for the the `keycloak.yml` and also not add anything to the build). We then follow through to the end to see if also our functional tests work as expected. I am also very curious to see if there are performance differences for ARM vs non-ARM. 

Once we have a full run and are sure that there are no blocker, and the results on performance, we can have another issue to also support mixed ARM clusters.

BTW, the nice thing about a pure ARM cluster is that also ISPN deploys on ARM, and we don't need to configure anything over there. And we also don't need to configure anything to ensure the operators run on ARM which would require additional tweaking.

So TL;DR: Let's stay with pure ARM as much as possible, and do the minimal changes necessary to do that until we get a full performance run with some numbers.

**Comment by kami619** [comment]
@ahus1 Here is the performance run against the ARM worker node cluster - https://github.com/kami619/keycloak-benchmark/blob/result-data/rosa_scalability/2024/11/01/result-20241101-152055-8a5d6450-113e-4bcf-94ed-10759ffcb3bb.json,

and the equivalent run against the X86 worker node cluster - https://github.com/keycloak/keycloak-benchmark/blob/result-data/rosa_scalability/2024/11/01/result-20241101-094728-5d7258ed-b448-4611-9b5c-3f34962e3318.json for comparison. 

Overall I think the ARM Graviton nodes had better performance, it would be worthwhile to merge the PR and continue to observe the performance difference using Horreum for a comprehensive view.

**Comment by mhajas** [comment]
It seems it is ok in nightly run:
https://github.com/keycloak/keycloak-benchmark/blob/969229a478ed90fc7977e9992d5fab5cd6d0d0a4/rosa_scalability/2024/11/04/result-20241104-063242-fa557b40-5318-43f8-9c9a-5c9b48256ae5.json#L358-L362

**Comment by ahus1** [comment]
Please do a sanity check of the first nightly run, then close this issue.

**Comment by ahus1** [comment]
Tentative next step: 
* Micro-Benchmark on Hashing Passwords (default alg on ARM vs. x64)
* Micro-Benchmark on token signing (default alg on ARM vs. x64)
* Micro-Benchmark on different token signing algorithms (which ones perform better on ARM?)

**Comment by kami619** [comment]
[CPU_Processor_family_benchmarks.zip](https://github.com/user-attachments/files/17836940/CPU_Processor_family_benchmarks.zip)

<h3><strong>Performance Metrics by Compute Machine Type</strong></h3>

Compute Machine Type | User Logins per Second per 1 vCPU per Pod | Credential Grants per Second per 1 vCPU
-- | -- | --
m7a.2xlarge | 140 | 1,372
m7i.2xlarge | 123 | 1,176
m7g.2xlarge | 102 | 737
c7g.2xlarge | 101 | 700




The tables summarizing the **50th percentile (median) response times** (in milliseconds) for different endpoints across the four compute machine types.

---

### **Memory Usage Test**

| Compute Machine Type | Browser to Log In Endpoint | Browser Posts Correct Credentials | Exchange Code |
|----------------------|----------------------------|-----------------------------------|---------------|
| **m7a.2xlarge**      | 32 ms                      | 78 ms                             | 25 ms         |
| **m7i.2xlarge**      | 38 ms                      | 89 ms                             | 30 ms         |
| **m7g.2xlarge**      | 32 ms                      | 86 ms                             | 27 ms         |
| **c7g.2xlarge**      | 37 ms                      | 91 ms                             | 29 ms         |

---

### **CPU Usage for Logins Test**

| Compute Machine Type | Browser to Log In Endpoint | Browser Posts Correct Credentials | Exchange Code | Browser Logout |
|----------------------|----------------------------|-----------------------------------|---------------|----------------|
| **m7a.2xlarge**      | 31 ms                      | 76 ms                             | 24 ms         | 13 ms          |
| **m7i.2xlarge**      | 37 ms                      | 87 ms                             | 28 ms         | 17 ms          |
| **m7g.2xlarge**      | 32 ms                      | 86 ms                             | 25 ms         | 12 ms          |
| **c7g.2xlarge**      | 36 ms                      | 91 ms                             | 29 ms         | 16 ms          |

---

### **CPU Usage for Credential Grants Test**

| Compute Machine Type | Client Credentials Grant Type |
|----------------------|-------------------------------|
| **m7a.2xlarge**      | 31 ms                         |
| **m7i.2xlarge**      | 36 ms                         |
| **m7g.2xlarge**      | 30 ms                         |
| **c7g.2xlarge**      | 36 ms                         |

---

**Comment by ahus1** [comment]
@kami619 - with the discussions we had, and that we will only do runs on `c7g.2xlarge` and we consider the results good enough, I assume this issue can be closed?

**Comment by kami619** [comment]
@ahus1 thats correct Alex, I am closing this task out now.
